test/test_binary_ufuncs.py

```diff
@@ -1382,112 +1373,74 @@ def test_fmod_by_zero_integral(self, device, dtype):
                 x.fmod(zero)
-            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
-                torch.fmod(x, zero, out=out)
-            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
-                x.fmod_(zero)
-        # Different value for different dtype on GPU
+        # Different value for different dtype on CUDA:
+        # Due to it's an undefined behavior, CUDA returns a pattern of all 1s
+        # for integral dividend (other than int64) divided by zero. For int64,
+        # CUDA returns all 1s for negative dividend, half 1s for positive dividend.
+        # uint8: 0xff -> 255
+        # int32: 0xffffffff -> -1
         else:
             if dtype == torch.int64:
                 self.assertEqual(x.fmod(zero) == 4294967295, x >= 0)
                 self.assertEqual(x.fmod(zero) == -1, x < 0)
-                # out
-                torch.fmod(x, zero, out=out)
-                self.assertEqual(out == 4294967295, x >= 0)
-                self.assertEqual(out == -1, x < 0)
-                self.assertEqual(out.size(), torch.Size([10, 10]))
-                # in-place
-                x_.fmod_(zero)
-                self.assertEqual(x_ == 4294967295, x >= 0)
-                self.assertEqual(x_ == -1, x < 0)
             else:
                 value = 255 if dtype == torch.uint8 else -1
                 self.assertTrue(torch.all(x.fmod(zero) == value))
-                # out
-                torch.fmod(x, zero, out=out)
-                self.assertTrue(torch.all(out == value))
-                self.assertEqual(out.size(), torch.Size([10, 10]))
-                # in-place
-                x_.fmod_(zero)
-                self.assertTrue(torch.all(x_ == value))
 
     @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False, include_bool=False, include_complex=False))
     def test_fmod(self, device, dtype):
         # Use numpy as reference
-        def _reference_implementation(x, mod):
+        def _helper(x, mod):
             np_x = x.cpu().numpy()
-            np_mod = 0
-            # No type promotion
-            # Issue #47779: https://github.com/pytorch/pytorch/issues/47779
-            if torch.is_tensor(mod):
-                np_mod = mod.cpu().numpy()
-            else:
-                np_mod = mod
-                # Non XLA platform needs to cast to int
-                if dtype in torch.testing.get_all_int_dtypes() and self.device_type in ['cpu', 'cuda']:
-                    np_mod = int(np_mod)
+            np_mod = mod.cpu().numpy() if torch.is_tensor(mod) else mod
             exp = np.fmod(np_x, np_mod)
             exp = torch.from_numpy(exp)
 
             res = torch.fmod(x, mod)
-            res = res.to(exp.dtype)
-            self.assertEqual(res, exp)
+            # Different Type promotion between numpy and torch for
+            # integral tensor with float scalar/tensor
+            dtype_flag = \
```

Comment:
See analogous comments on the other PR in this stack for how this test can be further refined.

Reply:
I will introduce another wrapper for all binary functions with type promotion.